src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp

```diff
@@ -559,6 +559,51 @@ struct CopyFileObjectStorageOperation final : public IDiskObjectStorageOperation
     }
 };
 
+struct TruncateFileObjectStorageOperation final : public IDiskObjectStorageOperation
+{
+    std::string path;
+    size_t size;
+
+    TruncateFileOperationOutcomePtr truncate_outcome;
+
+    TruncateFileObjectStorageOperation(
+        IObjectStorage & object_storage_,
+        IMetadataStorage & metadata_storage_,
+        const std::string & path_,
+        size_t size_)
+        : IDiskObjectStorageOperation(object_storage_, metadata_storage_)
+        , path(path_)
+        , size(size_)
+    {}
+
+    std::string getInfoForLog() const override
+    {
+        return fmt::format("TruncateFileObjectStorageOperation (path: {}, size: {})", path, size);
+    }
+
+    void execute(MetadataTransactionPtr tx) override
+    {
+        if (metadata_storage.exists(path))
+        {
+            if (!metadata_storage.isFile(path))
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Path {} is not a file", path);
+
+            truncate_outcome = tx->truncateFile(path,size);
+        }
+    }
+
+    void undo() override
+    {
+
+    }
```

Comment:
why not undo?
```suggestion
    {
       if (truncate_outcome)
           tx->undo();
    }
```

Reply:
Because there is no `undo` method in `MetadataStorageFromDiskTransaction`. It seems like it uses `add`, `commit` approach, so we can't do `undo` or `finalise`.

Reply:
Hmm, yeah, I was looking at https://github.com/ClickHouse/ClickHouse/blob/ec4c07d8100000532a03bfda48cb2d2169d11fd0/src/Disks/ObjectStorages/IMetadataOperation.h#L9-L15
when writing this comment.